File: tabela1.r
pacman::p_load(sdcMicro, readxl, dplyr, readr) 

dados <- read_delim(file.choose(), delim = ";", locale = locale(encoding = "Latin1")
)

dados <- dados |> select(
  -txt_nome_empreendimento,
  -cod_ibge,
  -cod_empreendimento
)

# as.Date transforma strings em datas, e "format" as transforma em ano-mês
# p.19, p.31, p.32
dados$data_referencia <- format(as.Date(dados$data_referencia), "%Y-%m")

dados$dt_assinatura_contrato <- format(as.Date(dados$dt_assinatura_contrato), "%Y-%m")

# p. 19 (agrupamento de idade) e p. 35 (generalização para faixas)
dados$faixa_etaria <- cut(as.numeric(format(as.Date(dados$data_nascimento), "%Y")),
                          breaks = c(1900, 1980, 2000, 2010, 2025),
                          labels = c("até 29", "30-49", "50-64", "65+"),
                          right = FALSE)

dados <- dados |> select(-data_nascimento)

colnames(dados)[colnames(dados) == "Estado Civil"] <- "estado_civil"
colnames(dados)[colnames(dados) == "Tipo de Beneficiário"] <- "tipo_beneficiario"

quasi_identificadores <- c(
  "txt_municipio", "txt_uf", "txt_regiao",
  "txt_sexo", "estado_civil", "tipo_beneficiario", "faixa_etaria"
)

sdc <- createSdcObj(
  dat = dados,
  keyVars = quasi_identificadores
)

sdc <- localSuppression(sdc, k = 5)

sdc <- pram(sdc, variables = c("estado_civil", "txt_sexo"))


tryCatch({
  dados_anonimizados <- extractManipData(sdc)
  write.csv(dados_anonimizados, "tabela1_anonimizada.csv", row.names = FALSE)
  print("Dados salvos!")
}, error = function(e) {
  message("Erro ao extrair os dados anonimizados. Verifique se o objeto 'sdc' foi criado corretamente.")
})

File: tabela1.py
import pandas as pd

#Ler base
dados = pd.read_csv("dados ficticios.csv", sep=";", encoding="latin1")

#Remover colunas irrelevantes
dados = dados.drop(columns=["txt_nome_empreendimento", "cod_ibge", "cod_empreendimento"], errors="ignore")

#Converter datas para ano-mês
dados["data_referencia"] = pd.to_datetime(dados["data_referencia"], dayfirst=True, errors="coerce").dt.to_period("M").astype(str)
dados["dt_assinatura_contrato"] = pd.to_datetime(dados["dt_assinatura_contrato"], dayfirst=True, errors="coerce").dt.to_period("M").astype(str)

#Criar faixa etária
dados["faixa_etaria"] = pd.to_datetime(dados["data_nascimento"], dayfirst=True, errors="coerce").dt.year
dados["faixa_etaria"] = pd.cut(dados["faixa_etaria"], 
                               bins=[1900, 1980, 2000, 2010, 2025], 
                               labels=["até 29", "30-49", "50-64", "65+"], 
                               right=False)

#Remover coluna original de nascimento
dados = dados.drop(columns=["data_nascimento"], errors="ignore")

#Renomear colunas
dados = dados.rename(columns={
    "Estado Civil": "estado_civil",
    "Tipo de Beneficiário": "tipo_beneficiario"
})

#Converter colunas monetárias para float
colunas_monetarias = ["vr_garantia_inicial", "vr_subsidio_concessao", "vr_renda_familiar_comprovada"]
for col in colunas_monetarias:
    if col in dados.columns:
        dados[col] = dados[col].astype(str).replace({"R\$": "", "\.": "", ",": "."}, regex=True)
        dados[col] = pd.to_numeric(dados[col], errors="coerce")

#Amostragem estratificada por modalidade
frac_por_grupo = 0.05  # 5% de cada modalidade
amostra = dados.groupby("txt_modalidade", group_keys=False).apply(
    lambda x: x.sample(frac=frac_por_grupo, random_state=123)
)

#Salvar amostra para usar no sdcApp
amostra.to_csv("amostra_sdc.csv", index=False, encoding="utf-8-sig")
print("✅ Amostra estratificada salva como 'amostra_sdc.csv'")

# EDA básica da amostra
print("\nEstrutura da amostra:")
print(amostra.info())

print("\nResumo estatístico das variáveis numéricas:")
print(amostra.describe(include="number"))

print("\nResumo estatístico das variáveis categóricas:")
print(amostra.describe(include="object"))

print("\nContagem por UF:")
print(amostra["txt_uf"].value_counts())

print("\nContagem por modalidade:")
print(amostra["txt_modalidade"].value_counts())
